XenAdmin/Wizards/ConversionWizard/SrSelectionPage.cs

```diff
@@ -229,13 +234,22 @@ private void comboBoxSr_SelectedIndexChanged(object sender, EventArgs e)
             if (comboBoxSr.SelectedItem is SrWrapper wrapper)
             {
                 SelectedSR = wrapper.item;
-
                 if (wrapper.AvailableSpace >= _requiredDiskSize)
-                    tableLayoutPanelError.Visible = false;
+                    if (!SelectedSR.VdiCreationCanProceed(_requiredDiskSize))
+                        ShowError(string.Format(Messages.SR_DISKSIZE_EXCEEDS_DISK_MAX_SIZE, Util.DiskSizeString(SR.DISK_MAX_SIZE, 0)));
+                    else
+                    {
+                        tableLayoutPanelError.Visible = false;
+                        wrapper.CanUse = true;
+                    }
+                else if (Array.IndexOf(SelectedSR.GetSM().capabilities, "LARGE_VDI") != -1)
+                    {
+                        ShowWarning(Messages.CONVERSION_STORAGE_PAGE_SR_OVERCOMMIT);
+                        wrapper.CanUse = true;
+                    }

```

Comment:
In the else-if you probably mean to check for `THIN_PROVISIONING` rather than `LARGE_VDI`,right?

I find this logic slightly more complex, plus that SelectedSR.VdiCreationCanProceed will run the thin provisioning check unnecessarily since it's nested under the `wrapper.AvailableSpace >= _requiredDiskSize` case.

One more thing that I'm thinking now is that, using VdiCreationCanProceed as is, is probably incorrect because it runs the thin provisioning check on the SR's free space, not the available space as calculated by the conversion manager. How about this one:

```
SM sm = SelectedSR.GetSM();
bool vdiSizeUnlimited = sm != null && Array.IndexOf(sm.capabilities, "LARGE_VDI") != -1;
bool isThinlyProvisioned = sm != null && Array.IndexOf(sm.capabilities, "THIN_PROVISIONING") != -1;

if (_requiredDiskSize > SR.DISK_MAX_SIZE && !vdiSizeUnlimited)
{
    ShowError(string.Format(Messages.SR_DISKSIZE_EXCEEDS_DISK_MAX_SIZE, Util.DiskSizeString(SR.DISK_MAX_SIZE, 0)));
    wrapper.CanUse = false;
}
else if (_requiredDiskSize > wrapper.AvailableSpace && !isThinlyProvisioned)
{
    ShowError(Messages.CONVERSION_STORAGE_PAGE_SR_TOO_SMALL);
    wrapper.CanUse = false;
}
else if (_requiredDiskSize > wrapper.AvailableSpace)
{
    ShowWarning(Messages.CONVERSION_STORAGE_PAGE_SR_OVERCOMMIT);
    wrapper.CanUse = true;
}
else
{
    tableLayoutPanelError.Visible = false;
    wrapper.CanUse = true;
}
```

Reply:
Thanks Konstantina, this logic is better clear. And I mean for  'overcommit' situation, we need to check both 'vdiSizeUnlimited' and 'isThinlyProvisioned'.

Reply:
Why do we need to check for vdiSizeUnlimited? I'm not aware of this requirement and indeed the corresponding code is SR.VdiCreationCanProceed does not related the two.

Reply:
This is because in xcm when convert vm >2TB, we can only support gfs2 storage(vdiSizeUnlimited and isThinlyProvisioned) now. Besides the function SR.VdiCreationCanProceed can't be used to decide whether a SR can be overcommitted. For example, SR with availableSpace=100G, requiredDiskSize=200G:
For SR like 'smb' ,'ext' and 'nfs', vdiSizeUnlimited= false , isThinlyProvisioned= true and  SR.VdiCreationCanProceed return true, but it'll fail in vdi create process in a convert job actually. So I need to both check for vdiSizeUnlimited and isThinlyProvisioned in overcommitted situation.

Reply:
In the example you have mentioned with availableSpace=100G, requiredDiskSize=200G, vdiSizeUnlimited= false , isThinlyProvisioned= true, you end up in the third block of the logic I have written above, which is the warning for overcommit. LARGE_VDI is not relevant, it only says whether the SR supports disk size >2TB.

Reply:
Year, what I want to say is, the overcommit situation is just a warning, the user can continue to submit this job and the job will success. But if we in the example above, I'm sure the job will fail. So I think it should go to Error.CONVERSION_STORAGE_PAGE_SR_TOO_SMALL and I change this logic to now. 
               if (_requiredDiskSize > SR.DISK_MAX_SIZE && !vdiSizeUnlimited)
                {
                    ShowError(string.Format(Messages.SR_DISKSIZE_EXCEEDS_DISK_MAX_SIZE, Util.DiskSizeString(SR.DISK_MAX_SIZE, 0)));
                }
               else if (_requiredDiskSize > wrapper.AvailableSpace && isThinlyProvisioned && vdiSizeUnlimited)
                {
                    ShowWarning(Messages.CONVERSION_STORAGE_PAGE_SR_OVERCOMMIT);
                    wrapper.CanUse = true;
                }
                else if (_requiredDiskSize > wrapper.AvailableSpace)
                {
                    ShowError(Messages.CONVERSION_STORAGE_PAGE_SR_TOO_SMALL);
                }
                else
                {
                    tableLayoutPanelError.Visible = false;
                    wrapper.CanUse = true;
                }

Reply:
Have you actually tested that the job will fail? And why is XCM different from the new disk creation?

Reply:
Oh, I find I make a mistake here. I always thought about we should allow to convert a VM more than 2TB, so I add the vdiSizeUnlimited check in overcommit. You are right, in normal case, just need to check isThinlyProvisioned. I add all the test results in tickets.